kernel/mmanutils.c

```diff
@@ -308,3 +517,140 @@ int myst_release_process_mappings(pid_t pid)
 done:
     return ret;
 }
+
+static int _sync_file(int fd, off_t offset, const void* addr, size_t length)
+{
+    int ret = 0;
+    const uint8_t* p = (const uint8_t*)addr;
+    size_t r = length;
+    off_t o = offset;
+
+    while (r > 0)
+    {
+        ssize_t n = pwrite(fd, p, r, o);
+
+        if (n == 0)
+            break;
+        else if (n < 0)
+            ERAISE(n);
+
+        p += n;
+        o += n;
+        r -= (size_t)n;
+    }
+
+done:
+    return ret;
+}
+
+int myst_msync(void* addr, size_t length, int flags)
+{
+    int ret = 0;
+    const int mask = MS_SYNC | MS_ASYNC | MS_INVALIDATE;
+
+    /* reject bad parameters and unknown flags */
+    if (!addr || !length || (flags & ~mask))
+        ERAISE(-EINVAL);
+
+    /* fail if both MS_SYNC and MS_ASYNC are both present */
+    if ((flags & MS_SYNC) && (flags & MS_ASYNC))
+        ERAISE(-EINVAL);
+
+    // Note: Asynchronous sync (MS_ASYNC) is not supported so all syncs are
+    // treated as though they were non-asynchronous (MS_SYNC). The caller will
+    // be unable to detect any difference in behavior.
+
+    // Note: experimentation reveals that Linux invalidates other mappings
+    // of the same file, whether  MS_INVALIDATE is present or not (meaning
+    // that they too are updated to reflect the contents of the file with
+    // or without the MS_INVALIDATE flag).
+
+    myst_spin_lock(&_msync_mappings_lock);
+    {
+        /* flush any msync mappings contained by this address range */
+        for (msync_mapping_t* p = _msync_mappings; p; p = p->next)
+        {
+            uint8_t* lo = addr;
+            uint8_t* hi = (uint8_t*)addr + length;
+            uint8_t* plo = p->addr;
+            uint8_t* phi = (uint8_t*)p->addr + p->length;
+            uint8_t* maxlo = _max_ptr(lo, plo);
+            uint8_t* minhi = _min_ptr(hi, phi);
+
+            // [AAAAAAAAAAAAAAAAA]
+            //       [MMMMMMMMMMMMMMMMMMMMM]
+            //       ^           ^
+            //      maxlo      minhi
+            //
+            //       [AAAAAAAAAAAAAAAAA]
+            // [MMMMMMMMMMMMMMM]
+            //       ^         ^
+            //      maxlo     minhi
+
+            if (maxlo < minhi)
+            {
+                ECHECK(_sync_file(
+                    p->fd,                     /* fd */
+                    p->offset + (maxlo - plo), /* offset */
+                    maxlo,                     /* addr */
+                    minhi - maxlo));           /* length */
+            }
```

Comment:
Could we have two FDs mmaped to overlapping memory regions? If not, we can break out of the loop here.

Reply:
Yes, you can map two contiguous regions of memory, map different files to each and then, and then free them as one.